File: clojure/06-io-operations/file_read.clj
;; Reading files in Clojure
(require '[clojure.java.io :as io])

;; Set up a sample file to read
(spit "poem.txt" "roses are red\nviolets are blue\nClojure is\nfunctional too\n")

;; slurp reads the whole file into one string
(println "Whole file with slurp:")
(println (slurp "poem.txt"))

;; For large files, stream line-by-line with a lazy sequence.
;; with-open closes the reader automatically.
(println "Line by line:")
(with-open [rdr (io/reader "poem.txt")]
  (doseq [line (line-seq rdr)]
    (println (str "> " line))))

;; line-seq composes with any sequence function, like count
(with-open [rdr (io/reader "poem.txt")]
  (println "Total lines:" (count (line-seq rdr))))

File: ruby/06-io-operations/file_read.rb
# Read the entire file into one string
contents = File.read("gemstones.txt")
puts "The file has #{contents.length} characters"

# Read into an array of lines, stripping newlines
lines = File.readlines("gemstones.txt", chomp: true)
puts "First gemstone: #{lines.first}"
puts "Total gemstones: #{lines.length}"

# Stream line by line without loading the whole file
File.foreach("gemstones.txt").with_index(1) do |line, number|
  puts "#{number}: #{line.chomp}"
end
